tests/numerics/ad/test_forward_mode.py

```diff
@@ -137,37 +153,41 @@ def test_advar_m_mul_vec_n():
     assert np.sum(z.jac.toarray() - Jz) == 0
 
 
-def test_mul_sps_advar():
-    J = sps.csc_matrix(np.array([[1, 3, 1], [5, 0, 0], [5, 1, 2]]))
+def test_mul_sps_advar(create_csc: type[sps.csc_matrix] | type[sps.csc_array]):
+    J = create_csc(np.array([[1, 3, 1], [5, 0, 0], [5, 1, 2]]))
     x = AdArray(np.array([1, 2, 3]), J)
-    A = sps.csc_matrix(np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
+    A = create_csc(np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
     f = A @ x
 
     assert np.all(f.val == [14, 32, 50])
-    assert np.all(f.jac == A * J.toarray())
+    assert np.all(f.jac == A @ J.toarray())
 
 
-def test_mul_advar_vectors():
-    Ja = sps.csc_matrix(np.array([[1, 3, 1], [5, 0, 0], [5, 1, 2]]))
-    Jb = sps.csc_matrix(np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]]))
+def test_mul_advar_vectors(create_csc: type[sps.csc_matrix] | type[sps.csc_array]):
+    Ja = create_csc(np.array([[1, 3, 1], [5, 0, 0], [5, 1, 2]]))
+    Jb = create_csc(np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]]))
     a = AdArray(np.array([1, 2, 3]), Ja)
     b = AdArray(np.array([1, 1, 1]), Jb)
-    A = sps.csc_matrix(np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
+    A = create_csc(np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
 
     f = A @ a + b
 
     assert np.all(f.val == [15, 33, 51])
-    assert np.sum(f.jac.toarray() != A * Ja + Jb) == 0
-    assert (
-        np.sum(Ja != sps.csc_matrix(np.array([[1, 3, 1], [5, 0, 0], [5, 1, 2]]))) == 0
-    )
-    assert (
-        np.sum(Jb != sps.csc_matrix(np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]]))) == 0
-    )
+    assert compare_matrices(f.jac, A @ Ja + Jb)
+    # Asterix stands for element-wise multiplication for sparrays and matrix
+    # multiplication for spmatrix.
+    if create_csc is sps.csc_matrix:
+        assert compare_matrices(f.jac, A * Ja + Jb)
+    elif create_csc is sps.csc_array:
+        assert not compare_matrices(f.jac, A * Ja + Jb)
```

Comment:
Why do csc matrix and array behave differently?

Reply:
For all the use cases within the ad framework - no it does not, as the ad operator overrides for @ and * remain the same as before (consistent to the arrays). The cases when an array and not a matrix is passed are handled and tested, also in this file.

Outside the ad framework - yes, one will face troubles if using them interchangeably. As I see, the dangerous part is everything related to discretizations, but these issues should be early discovered in their own tests

Reply:
Yes